modules/ecs-cluster-ec2-asg/v0.0.1/main.tf

```diff
@@ -0,0 +1,308 @@
+
+  metadata_options {
+    # Enable Instance Metadata Service V2 
+    http_tokens = "required"
+  }
+
+  iam_instance_profile {
+    name = aws_iam_instance_profile.ecs_instance_profile.id
+  }
+
+  monitoring {
+    enabled = true
+  }
+
+  tag_specifications {
+    resource_type = "network-interface"
+    tags          = var.tags
+  }
+
+  tag_specifications {
+    resource_type = "volume"
+    tags          = var.tags
+  }
+}
+
+# Create Auto-Scaling Group
+resource "aws_autoscaling_group" "this" {
+  name                      = var.cluster_name
+  vpc_zone_identifier       = var.subnet_ids
+  min_size                  = var.min_size
+  max_size                  = var.max_size
+  desired_capacity          = var.desired_capacity
+  health_check_type         = var.health_check_type
+  health_check_grace_period = var.health_check_grace_period
+  default_cooldown          = var.default_cooldown
+  termination_policies      = var.termination_policies
+  protect_from_scale_in     = var.protect_from_scale_in
+  # protect_from_scale_in     = true
+  # To enable managed termination protection for a capacity provider, the Auto Scaling group must have instance protection from scale in enabled.
+
+  launch_template {
+    id      = aws_launch_template.this.id
+    version = "$Latest"
+  }
+
+  tag {
+    key                 = "ecs_cluster"
+    value               = var.cluster_name
+    propagate_at_launch = true
+  }
+
+  dynamic "tag" {
+    for_each = var.tags
+
+    content {
+      key                 = tag.key
+      value               = tag.value
+      propagate_at_launch = true
+    }
+  }
+
+
+  lifecycle {
+    create_before_destroy = true
+  }
+}
+
+# Create capacity provider
+# Reference: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/cluster-capacity-providers.html
+resource "aws_ecs_capacity_provider" "this" {
+  #The capacity provider name can have up to 255 characters, including letters (upper and lowercase), numbers, underscores, and hyphens.
+  # The name cannot be prefixed with "aws", "ecs", or "fargate".
+  name = join("-", ["cp", var.cluster_name])
+
+  auto_scaling_group_provider {
+    auto_scaling_group_arn = aws_autoscaling_group.this.arn
+    # managed_termination_protection = "ENABLED"
+    managed_termination_protection = "DISABLED"
+
+    managed_scaling {
+      instance_warmup_period    = 120 # Default is 300
+      maximum_scaling_step_size = 1000
+      minimum_scaling_step_size = 1
+      status                    = "ENABLED"
+      target_capacity           = 100
+    }
+  }
+}
+
+resource "aws_ecs_cluster_capacity_providers" "this" {
+  cluster_name = aws_ecs_cluster.this.name
+
+  capacity_providers = [aws_ecs_capacity_provider.this.name]
+
+  default_capacity_provider_strategy {
+    base              = 1
+    weight            = 100
+    capacity_provider = aws_ecs_capacity_provider.this.name
+  }
+}
+
+# Create autoscaling policies
+resource "aws_autoscaling_policy" "up" {
```

Comment:
i think ECS cap providers cannot work together with other scaling policies. That might be why you saw the flip flopping.

Reply:
These were removed (disabled) when testing the capacity providers, and a re-deploy was done when adding the capacity providers so they didn't exist.